### PR TITLE
Implement PEG from ICLR 2022

### PR DESCRIPTION
**Pull request by zoom-wang112358**
[PEG](https://openreview.net/pdf?id=e95i1IHcWj) uses a GCN layer with edge weights according to the distance between the end nodes of the edge and keep the positional encodings unchanged. The positional encodings could to be calculated by some graph embedding techniques, such as DeepWalk and Laplacian eigenmap.

Examples and descriptions could be found [here](https://github.com/Graph-COM/PEG/tree/main/PEG_layer).

### CODE REVIEW

File: torch_geometric/nn/conv/peg_conv.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            symmetric normalization coefficients on the fly.
+            (default: :obj:`True`)
+        bias (bool, optional): If set to :obj:`False`, the layer will not learn
+            an additive bias. (default: :obj:`True`)
+        use_formerinfo (bool):
+            Whether to use previous layer's output to update node features.
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+    def __init__(self, in_feats_dim: int, pos_dim: int, out_feats_dim: int,
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:49]:
```suggestion
    def __init__(self, in_channels: int, pos_dim: int, out_channels: int,
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        self.edge_mlp_dim = edge_mlp_dim
+
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+        self.weight_withformer = Parameter(
+            torch.Tensor(in_feats_dim + in_feats_dim, out_feats_dim))
+        self.weight_noformer = Parameter(
+            torch.Tensor(in_feats_dim, out_feats_dim))
+        self.edge_mlp = nn.Sequential(nn.Linear(1, edge_mlp_dim),
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:75]: [newbie question] somewhat confused by this - If you go from 1 dim back to 1 dim are you able learning much?

File: torch_geometric/nn/conv/peg_conv.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            hidden_out = torch.cat([hidden_feats, m_i], dim=-1)
+            hidden_out = hidden_out @ self.weight_withformer
+        else:
+            hidden_out = m_i
+            hidden_out = hidden_out @ self.weight_noformer
+
+        # return tuple
+        return self.update((hidden_out, coors_out), **update_kwargs)
+
+    def glorot(self, tensor):
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:184]: available in `torch_geometric.nn.inits`

File: torch_geometric/nn/conv/peg_conv.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        if tensor is not None:
+            stdv = math.sqrt(6.0 / (tensor.size(-2) + tensor.size(-1)))
+            tensor.data.uniform_(-stdv, stdv)
+
+    def zeros(self, tensor):
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:189]: same as above

File: torch_geometric/nn/conv/peg_conv.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                pos) -> Tensor:
+        PE_edge_weight = self.edge_mlp(pos)
+        return x_j if edge_weight is None else (PE_edge_weight *
+                                                edge_weight.view(-1, 1) * x_j)
+
+    def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
+        r"""The initial call to start propagating messages.
+
+        Args:
+            `edge_index` holds the indices of a general (sparse)
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:153]:
```suggestion
            edge_index (Tensor): holds the indices of a general (sparse)
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                assignment matrix of shape :obj:`[N, M]`.
+            size (tuple, optional) if none, the size will be inferred
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:155]:
```suggestion
            size (tuple, optional): if none, the size will be inferred
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+        return torch.cat([coors_out, hidden_out], dim=-1)
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_weight: OptTensor,
+                pos) -> Tensor:
+        PE_edge_weight = self.edge_mlp(pos)
+        return x_j if edge_weight is None else (PE_edge_weight *
+                                                edge_weight.view(-1, 1) * x_j)
+
+    def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
```
Comment (Padarn) [inline torch_geometric/nn/conv/peg_conv.py:149]: Sorry I didn't fully follow the implementation yet - but I think you should be able to do this without needing to reimplement `propagate` - I.e you can put most of this logic inside your forward function.

File: torch_geometric/nn/conv/peg_conv.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.nn.conv.gcn_conv import gcn_norm
+from torch_geometric.typing import Adj, OptTensor, Size, Tensor
+
+
+class PEGConv(MessagePassing):
+    r"""The PEG layer from the
+    `"Equivariant and Stable Positional Encoding
+    for More Powerful Graph Neural Networks"
+    <https://arxiv.org/abs/2203.00199>`_ paper
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:17]: It would be great if we can add more description here about the underlying operator :)

File: torch_geometric/nn/conv/peg_conv.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+
+    Args:
+        in_feats_dim (int): Size of input node features.
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:21]: Align with PyG naming convention:
```suggestion
        in_channels (int): Size of input node features.
        out_channels (int): ...
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        self.improved = improved
+        self.cached = cached
+        self.add_self_loops = add_self_loops
+        self.normalize = normalize
+        self.edge_mlp_dim = edge_mlp_dim
+
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+        self.weight_withformer = Parameter(
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:71]: Looks like these parameters are never properly initialized.

File: torch_geometric/nn/conv/peg_conv.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        self.glorot(self.weight_noformer)
+        self.zeros(self.bias)
+        self._cached_edge_index = None
+        self._cached_adj_t = None
+
+    def forward(self, x: Tensor, edge_index: Adj,
+                edge_weight: OptTensor = None) -> Tensor:
+        """"""
+
+        coors, feats = x[:, :self.pos_dim], x[:, self.pos_dim:]
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:96]: Any reason we prefer slicing here over something like
```python
def forward(self, x, pos, edge_index, edge_weight):
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+                if cache is None:
+                    edge_index = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, feats.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_adj_t = edge_index
+                else:
+                    edge_index = cache
+        else:
+            print('We normalize the adjacency matrix in PEG.')
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:121]: Let's not do any printing here :)

File: torch_geometric/nn/conv/peg_conv.py (hunk 7 above)
Comment (rusty1s) [reply torch_geometric/nn/conv/peg_conv.py:149]: +1 . We definitely do not want to override `propagate` here. Any reason you did this in the first place?

File: torch_geometric/nn/conv/peg_conv.py (hunk 2 above)
Comment (ZoomWang666) [reply torch_geometric/nn/conv/peg_conv.py:75]: Maybe I can use 'information bottleneck' to explain it. Though I go from 1 dim back to 1 dim, the network rids noisy input data of extraneous details as if by squeezing the information through a bottleneck, retaining only the features most relevant to the edge weight.

File: torch_geometric/nn/conv/peg_conv.py (hunk 7 above)
Comment (ZoomWang666) [reply torch_geometric/nn/conv/peg_conv.py:149]: Thanks for your advice. I put the operational logic inside my forward function.

**Comment by ZoomWang666** [comment]
@rusty1s  Have just modified the implementation. Please review all the files and let me know if anything more is required. Thanks so much.

**Comment by rusty1s** [comment]
Thanks. I will take a last look soon :)

File: CHANGELOG.md [hunk 13]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [2.0.5] - 2022-MM-DD
 ### Added
+- Added the `PEG` model ([#4571](https://github.com/pyg-team/pytorch_geometric/pull/4571))
```
Comment (rusty1s) [inline CHANGELOG.md:8]:
```suggestion
- Added the `PEGConv` layer from the ["Equivariant and Stable Positional Encoding for More Powerful Graph Neural Networks"](https://arxiv.org/abs/2203.00199>) paper ([#4571](https://github.com/pyg-team/pytorch_geometric/pull/4571))
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+from torch_geometric.typing import Adj, OptTensor, Tensor
+
+
+class PEGConv(MessagePassing):
+    r"""The PEG layer from the
+    `"Equivariant and Stable Positional Encoding
+    for More Powerful Graph Neural Networks"
+    <https://arxiv.org/abs/2203.00199>`_ paper.
+
+    $$X^{'},Z^{'}=(\sigma [(\hat{A} \odot M)XW],Z)$$
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:19]: Can we follow the `.. math::` syntax here as in other layers? The formula also looks wrong: https://pytorch-geometric--4571.org.readthedocs.build/en/4571/modules/nn.html#torch_geometric.nn.conv.PEGConv

File: torch_geometric/nn/conv/peg_conv.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    where\ $M_{uv}=MLP(||Z_u-Z_v||),\forall u,v \in V$.
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:20]: Here again. Use
```
:math:`M_{u,v} ...`
```

File: torch_geometric/nn/conv/peg_conv.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+        return out
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_weight: OptTensor,
+                pos) -> Tensor:
+        PE_edge_weight = self.edge_mlp(pos)
+        return x_j if edge_weight is None else (PE_edge_weight *
+                                                edge_weight.view(-1, 1) * x_j)
+
+    def __repr__(self):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:147]: Can be dropped.

File: torch_geometric/nn/conv/peg_conv.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        out = self.lin(out)
+        if self.bias is not None:
+            out += self.bias
+
+        return out
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_weight: OptTensor,
+                pos) -> Tensor:
+        PE_edge_weight = self.edge_mlp(pos)
+        return x_j if edge_weight is None else (PE_edge_weight *
```
Comment (rusty1s) [inline torch_geometric/nn/conv/peg_conv.py:144]: Shouldn't we always use `edge_mlp` here?
```python
out = x_j * self.edge_mlp(pos)
return out if edge_weight is None else edge_weight.view(-1, 1) * out
```

**Review by rusty1s — COMMENTED** [review]
Thanks for the updates. This looks good. Please add your layer to the `README.md` as well.

**Comment by rusty1s** [comment]
Hi @ZoomWang666, thanks again for this PR. I would like to merge it once the open conversations are resolved. Can you take a look if you have time and let me know? Best
